cpp/src/arrow/compute/memory_resources.cc

```diff
@@ -0,0 +1,155 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/compute/memory_resources.h"
+
+#include <array>
+#include <memory>
+#include <mutex>
+#include <random>
+
+#include "arrow/compute/exec.h"
+#include "arrow/util/logging.h"
+
+#ifdef __APPLE__
+#include <sys/sysctl.h>
+#include <sys/types.h>
+#endif
+
+#ifdef __linux__
+#include <sys/statvfs.h>
+#include <sys/sysinfo.h>
+#endif
+
+// Windows APIs
+#include "arrow/util/windows_compatibility.h"
+
+namespace arrow {
+
+namespace compute {
+
+std::string MemoryLevelName(MemoryLevel memory_level) {
+  static const char* MemoryLevelNames[] = {ARROW_STRINGIFY(MemoryLevel::kDiskLevel),
+                                           ARROW_STRINGIFY(MemoryLevel::kCpuLevel),
+                                           ARROW_STRINGIFY(MemoryLevel::kGpuLevel)};
+
+  return MemoryLevelNames[static_cast<int>(memory_level)];
+}
+
+std::string MemoryResource::ToString() const { return MemoryLevelName(memory_level_); }
+
+MemoryResources::~MemoryResources() {}
+
+std::unique_ptr<MemoryResources> MemoryResources::Make() {
+  return std::unique_ptr<MemoryResources>(new MemoryResources());
+}
+
+Status MemoryResources::AddMemoryResource(std::shared_ptr<MemoryResource> resource) {
+  auto level = static_cast<size_t>(resource->memory_level());
+  if (stats_[level] != nullptr) {
+    return Status::KeyError("Already have a resource type registered with name: ",
+                            resource->ToString());
+  }
+  stats_[level] = std::move(resource);
+  return Status::OK();
+}
+
+size_t MemoryResources::size() const { return stats_.size(); }
+
+Result<MemoryResource*> MemoryResources::memory_resource(MemoryLevel memory_level) const {
+  auto level = static_cast<size_t>(memory_level);
+  if (stats_[level] == nullptr) {
+    return Status::KeyError("No memory resource registered with level: ",
+                            MemoryLevelName(memory_level));
+  }
+  return stats_[level].get();
+}
+
+std::vector<MemoryResource*> MemoryResources::memory_resources() const {
+  std::vector<MemoryResource*> arr;
+  for (auto&& resource : stats_) {
+    if (resource != nullptr) {
+      arr.push_back(resource.get());
+    }
+  }
+  return arr;
+}
+
+namespace {
+
+size_t GetTotalMemorySize() {
```

Comment:
Can you add a comment explaining what is being returned here?  Does this number include swap for example?  Or is it based on physical RAM.  Are there any caveats to be aware of?